### ISSUE TITLE
First call of Container.GetInstance creates two objects instead of one

### ISSUE THREAD

**Issue by OleksandrKrutykh** [simpleinjector/SimpleInjector#837]
### Description
After upgrading from SimpleInjector 4.10.2 to SimpleInjector 5.0.3, a first call of SimpleInjector.Container.GetInstance method for a class that was registered with Lifestyle.Transient or Lifestyle.Scoped creates two objects instead of one. Subsequent calls of Container.GetInstance method requesting an instance of the same class properly create a single object each. 

While creating a single extra object per class does not seem like a big deal, in my case it broke unit tests that verified some business logic by counting the number of objects created by the dependency injector. Also, such behavior may introduce bugs if the logic of a scoped class created through SimpleInjector assumes that there can't be more than one object of the class per an async scope.

## Expected behavior
The first call of Container.GetInstance creates a single instance of the requested class.

## Actual behavior
The first call of Container.GetInstance creates two objects. The second object is returned by GetInstance to the calling code. 

## To Reproduce
Here's a .NET Core 3.1 console app that reproduces the issue for a class that registered with Transient lifestyle. TestClass stores all created objects of the class inside TestClass.CreatedObjects list.

```csharp
using SimpleInjector;
using SimpleInjector.Lifestyles;
using System;
using System.Collections.Generic;

namespace ConsoleApp2
{
    public class TestClass
    {
        public TestClass() => CreatedObjects.Add(this);

        public bool DoSomethingWasCalled { get; private set; }
        public static List<TestClass> CreatedObjects { get; } = new List<TestClass>();

        public void DoSomething() => DoSomethingWasCalled = true;
    }

    class Program
    {
        static void Main()
        {
            var container = new Container
            {
                Options = { DefaultScopedLifestyle = new AsyncScopedLifestyle() }
            };

            container.Register<TestClass>(Lifestyle.Transient);

            var createdObject = container.GetInstance<TestClass>();
            createdObject.DoSomething();

            for (int i = 0; i < TestClass.CreatedObjects.Count; i++)
            {
                TestClass testObject = TestClass.CreatedObjects[i];
                Console.WriteLine($"Object #{i}: DoSomething was called = {testObject.DoSomethingWasCalled}");
            }
        }
    }
}
```

If the application references the SimpleInjector NuGet package of version 4.10.2, the application produces the following output:
```
Object #0: DoSomething was called = True
```

If SimpleInjector 5.0.3 is used, the output is the following:
```
Object #0: DoSomething was called = False
Object #1: DoSomething was called = True
```

**Comment by dotnetjunkie** [comment]
I'm closing this issue as a duplicate of [this one](https://github.com/simpleinjector/SimpleInjector/issues/834). Please let me know if the answer to that issue does not satisfy your needs.

**Comment by OleksandrKrutykh** [comment]
@dotnetjunkie Thank you very much for the answer! I set `EnableAutoVerification = false` option for the Container and it worked with SimpleInjector 5.0.3. The Contaner that the auto-verification was disabled for is the one used by specific unit tests, I did not disable auto-verification for my application itself. 

Would it make sense to add a remark about this behavior to the documentation? The [section about Scoped lifestyles](https://simpleinjector.readthedocs.io/en/latest/lifetimes.html#scoped-lifestyle) contains a table with phrases like "There will be only one instance of a given service..." or "Only one instance will be created...", which kind of contradicts the fact that there can be two instances of a service. I was thinking about something like this:

_Note: Although Simple Injector guarantees that only one instance of a service will be returned from the Container or injected through Auto-Wiring within a scope, under some circumstances Simple Injector can create additional instances of the service within the scope. These additional instances will not be visible to the application that uses Simple Injector and will be properly disposed at the end of the scope. Creation of additional service instances may be caused, for example, by auto-verification performed by the Container when the application requests a service from it for the first time._

A similar note could be added to the [section about Transient Lifestyle](https://simpleinjector.readthedocs.io/en/latest/lifetimes.html#transient-lifestyle):

_Note: Simple Injector can occasionally create more than one instance of a service when a service is requested from the Container. This can happen, for example, as a result of the Container performing auto-verification when a service is requested from it for the first time. These additional instances will not be returned to the calling code._

While I understand that these remarks look technical and may confuse a beginner who is not familiar with object lifestyles, they may be useful in reassuring a developer that this behavior is not a bug. My main concern upon discovering that Simple Injector creates extra service instances was that these instances would not be disposed and might therefore cause resource leaks. I was fortunate enough to notice that this behavior was caused by the version change of Simple Injector, so I simply rolled back to an earlier version. Some other developer may be reluctant to use Simple Injector at all after discovering such behavior.
